### PR TITLE
Add initial MinGW support

### PR DESCRIPTION
**Pull request by matwey**

### CODE REVIEW

**Comment by matwey** [comment]
@desowin could you have a look? I've managed to add CI tests for Windows. I would like to have CI tests for Visual Studio also.

**Comment by desowin** [comment]
This doesn't work for me. When I run the `load` program it crashes in `ov_open()` because `ov` is `NULL`. `ov_new()` returns NULL because it enters the path that sets `"Cannot find OV_FWPKG_RESOURCE resource"` error string (sadly, this string never reaches the caller, I single stepped that in debugger).

After `FindResource()` fails, the `GetLastError()` returns ERROR_RESOURCE_TYPE_NOT_FOUND.

**Comment by matwey** [comment]
Screenshot_windows_2019-10-23_18:56:08

For me, it works. What MinGW version do you use?

**Comment by desowin** [comment]
How do I tell which version I am using? I installed msys2-x86_64-20190524.exe then run `pacman -Syuu` twice and installed the packages using the commands from `.appveyor.yml`.

Could you please email me your build folder so we can determine if it is my system or my build setup that's causing issues?

File: src/bit.c [hunk 1]
```diff
@@ -5,7 +5,15 @@
 #include <stdint.h>
 #include <string.h>
 
-#include <arpa/inet.h>
+#if __BYTE_ORDER__ == __ORDER_LITTLE_ENDIAN__
+#	define be_to_host_16(x) (__builtin_bswap16(x))
```
Comment (desowin) [inline src/bit.c:9]: Visual Studio doesn't have __builtin_bswap16() nor __builtin_bswap32(). This is why I replaced the memcpy() + conversion with simple functions in https://github.com/matwey/libopenvizsla/pull/6/commits/23c8812de37211d2e6d7fb054262efe11a6ce1a1

Comment (matwey) [reply src/bit.c:9]: VS has `_byteswap_ushort` and `_byteswap_ulong` respectively.
I think there is no issue to use them here:
```
#if _MSC_VER
                _byteswap_ushort(x);
#else
                __builtin_bswap16(x);
#endif
```

**Comment by desowin** [comment]
I have checked the build you sent me. The difference is the fact that you didn't add `-DBUILD_SHARED_LIBS:BOOL=on` flag to CMake. I did add it as this is what is in `.appveyor.yml`.

When I build without shared libs *and* with the bugfixes merged, it works fine on my setup. The binary you sent me just froze during loading. I am using libusbK drivers.

**Comment by matwey** [comment]
> I have checked the build you sent me. The difference is the fact that you didn't add `-DBUILD_SHARED_LIBS:BOOL=on` flag to CMake. I did add it as this is what is in `.appveyor.yml`.
> 
> When I build without shared libs _and_ with the bugfixes merged, it works fine on my setup. The binary you sent me just froze during loading. I am using libusbK drivers.

Now I see the issue, thank you for testing! :-)

**Comment by matwey** [comment]
Hi,

I've fixed the issue and force-pushed the branch. Could you please retry?

**Review by desowin — APPROVED** [review]
With the latest changes, the executables find embedded fwpkg just fine. For the firmware to successfully load on my setup I had to merge the bugfixes (which are in master branch), so after this is merged the resulting binary will work just fine.

For some reason the ovextcap does not work with Wireshark. It fails with "Cannot open fifo for writing". I believe this is related to the msys standard library that doesn't correctly handle fopen() with pipes (the Visual Studio build works with Wireshark). When I run ovextcap with file as output, it works just fine.

**Comment by matwey** [comment]
Then I'll merge this pr.

Frankly speaking, I have not managed to try extcap on Linux yet.

**Comment by desowin** [comment]
> For some reason the ovextcap does not work with Wireshark. It fails with "Cannot open fifo for writing". I believe this is related to the msys standard library that doesn't correctly handle fopen() with pipes (the Visual Studio build works with Wireshark). When I run ovextcap with file as output, it works just fine.

I have created small code snippet that exhibits this discrepancy and submitted it to https://github.com/msys2/MSYS2-packages/issues/1769

**Comment by desowin** [comment]
> > For some reason the ovextcap does not work with Wireshark. It fails with "Cannot open fifo for writing". I believe this is related to the msys standard library that doesn't correctly handle fopen() with pipes (the Visual Studio build works with Wireshark). When I run ovextcap with file as output, it works just fine.
> 
> I have created small code snippet that exhibits this discrepancy and submitted it to [msys2/MSYS2-packages#1769](https://github.com/msys2/MSYS2-packages/issues/1769)

There is clear explaination to the mystery. The `ovextcap` works correctly if the `/mingw32/bin/gcc` or `/mingw64/bin/gcc` is used to compile it, and fails if `/usr/bin/gcc` is used. This is due to the later linking against `msys-2.0.dll` which is based on cygwin runtime that does not support named pipes.

AppVeyor is using `/mingw32/bin/gcc` and `/mingw64/bin/gcc` and thus the `ovextcap` built by AppVeyor should be working just fine.

**Pull request merged**